### PR TITLE
feat: support advanced matching based on post form

### LINKED ISSUE #5399: Proposal: support advanced matching based on POST form
# Proposal

## Background 

Currently, advanced matching only uses URL query data when matching on request content, and does not support advanced matching with fields in POST form as conditions.

## Objectives

Support for POST form as matching conditions by adding new matching key.

### Determine the added matching key prefix

I think we can add a new `post_arg_` prefix as a POST form data prefix, this is similar to the `arg_` used in the query parameter currently provided by openresty.

### Modify the mt metatable in `ctx.lua` and add a new key processor

The main purpose of the current modification is to handle POST form submissions with `Content-Type x-www-form-urlencoded`.

Use `ngx.req.read_body()` and fetch data from `ngx.req.get_post_args()` according to read requirements in the metatable to achieve complete on-demand inert loading with the help of the metatable feature. The currently used `lua-resty-expr` can automatically get the target value and compare it with the meta table based on the expression.

current `arg_`: 
[https://github.com/apache/apisix/blob/042ce5c1a6a4e040d6d6924f2d86e5534b730134/apisix/core/ctx.lua#L167-L176](https://github.com/apache/apisix/blob/042ce5c1a6a4e040d6d6924f2d86e5534b730134/apisix/core/ctx.lua#L167-L176)

demo: 
```lua
if core_str.has_prefix(key,"post_arg_") then
  ngx.req.read_body()
  local args, err = ngx.req.get_post_args()
  loacl post_arg_key = sub_str(key, 10)
  val = args[post_arg_key]
```

## Pre-fetching modifications

1. `apisix/core/ctx.lua`: Add new key processor
2. `t/*`: Add test case
3. Add documents for new function
4. Optional: Add support for its configuration to the Dashboard

## Other

What are your ideas?

### PR DESCRIPTION
**Pull request by bzp2010**
### What this PR does / why we need it:

Support for POST form as matching conditions in advanced matching.

Fix #5399 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: t/core/request.t [hunk 1]
```diff
@@ -378,3 +378,41 @@ nil
...
+
+
+=== TEST 11: get_post_args
+--- config
+    location = /hello {
+        real_ip_header X-Real-IP;
+
+        set_real_ip_from 0.0.0.0/0;
+        set_real_ip_from ::/0;
+        set_real_ip_from unix:;
```
Comment (spacewander) [inline t/core/request.t:391]: Do we need the real_ip stuff?

File: t/core/request.t [hunk 2]
```diff
@@ -378,3 +378,41 @@ nil
...
+            local ngx_ctx = ngx.ctx
+            local api_ctx = ngx_ctx.api_ctx
+            if api_ctx == nil then
+                api_ctx = core.tablepool.fetch("api_ctx", 0, 32)
+                ngx_ctx.api_ctx = api_ctx
+            end
+
+            core.ctx.set_vars_meta(api_ctx)
+        }
+        content_by_lua_block {
```
Comment (spacewander) [inline t/core/request.t:404]: Look like we can put the thing from access_by_lua_block into content_by_lua_block

File: t/node/vars.t [hunk 3]
```diff
@@ -299,3 +299,87 @@ GET /hello?k=uri_arg
...
+GET /hello
+--- error_code: 404
+--- response_body
+{"error_msg":"404 Route Not Found"}
+--- no_error_log
+[error]
+
+
+
+=== TEST 19: /not_found
```
Comment (spacewander) [inline t/node/vars.t:350]: Should use different names for consequential different tests

File: apisix/core/ctx.lua [hunk 4]
```diff
@@ -175,6 +175,20 @@ do
                     end
                 end
 
+            elseif core_str.has_prefix(key, "post_arg_") then
+                -- only match default post form
+                if request.header(nil, "Content-Type") == "application/x-www-form-urlencoded" then
```
Comment (spacewander) [inline apisix/core/ctx.lua:180]: Let's add a new section in https://github.com/apache/apisix/blob/master/docs/en/latest/router-radixtree.md

File: t/node/vars.t (hunk 3 above)
Comment (bzp2010) [reply t/node/vars.t]: changed

File: t/core/request.t (hunk 2 above)
Comment (bzp2010) [reply t/core/request.t]: merged

File: t/core/request.t (hunk 1 above)
Comment (bzp2010) [reply t/core/request.t]: removed

File: apisix/core/ctx.lua (hunk 4 above)
Comment (bzp2010) [reply apisix/core/ctx.lua:180]: @spacewander I've added a brief description, check to see if you need to add a separate paragraph `Howto` for descripte.

File: docs/en/latest/router-radixtree.md [hunk 5]
```diff
@@ -216,7 +217,7 @@ $ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f
 }'
 ```
 
-This route will require the request header `host` equal `iresty.com`, request cookie key `_device_id` equal `a66f0cdc4ba2df8c096f74c9110163a9` etc.
+This route will require the request header `host` equal `iresty.com`, request cookie key `_device_id` equal `a66f0cdc4ba2df8c096f74c9110163a9` etc. And the POST form property match requires you to use `Content-Type: application/x-www-form-urlencoded` for the request to work properly.
```
Comment (spacewander) [inline docs/en/latest/router-radixtree.md:220]: I think the post form property worths an individual section like the `graphql_` one.

Comment (bzp2010) [reply docs/en/latest/router-radixtree.md:220]: I will modify it.

File: docs/en/latest/router-radixtree.md [hunk 6]
```diff
@@ -218,6 +218,31 @@ $ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f
 
 This route will require the request header `host` equal `iresty.com`, request cookie key `_device_id` equal `a66f0cdc4ba2df8c096f74c9110163a9` etc.
 
+### How to filter route by POST form attributes {#post_args}
```
Comment (spacewander) [inline docs/en/latest/router-radixtree.md:221]: We don't need `{#post_args}` now. The anchor is not ambiguous.

Comment (bzp2010) [reply docs/en/latest/router-radixtree.md:221]: If the value of the anchor is not specified, a mismatch will occur between the English document and the Chinese document, is this really not necessary?

File: docs/en/latest/router-radixtree.md (hunk 5 above)
Comment (bzp2010) [reply docs/en/latest/router-radixtree.md]: fixed

File: docs/en/latest/router-radixtree.md (hunk 6 above)
Comment (spacewander) [reply docs/en/latest/router-radixtree.md:221]: Just using the name displayed in the title is enough. It's quite strange to use a custom anchor (and hard to remember too).

Comment (bzp2010) [reply docs/en/latest/router-radixtree.md]: OK, I've removed it.

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**